plugin/completion.py

```diff
@@ -1,36 +1,83 @@
 import sublime
 import sublime_plugin
-from .core.completion import parse_completion_response, format_completion
-from .core.configurations import is_supported_syntax
-from .core.documents import position_is_word
-from .core.edit import parse_text_edit
+
+from .core.protocol import Request, Range, InsertTextFormat
+from .core.settings import settings, client_configs
 from .core.logging import debug
-from .core.protocol import Request
+from .core.completion import parse_completion_response, format_completion
 from .core.registry import session_for_view, client_from_session, LSPViewEventListener
+from .core.configurations import is_supported_syntax
 from .core.sessions import Session
-from .core.settings import settings, client_configs
+from .core.edit import parse_text_edit
+from .core.views import range_to_region
 from .core.typing import Any, List, Dict, Tuple, Optional, Union
 from .core.views import text_document_position_params
 
 
-class CompletionState(object):
-    IDLE = 0
-    REQUESTING = 1
-    APPLYING = 2
-    CANCELLING = 3
+class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
+    def run(self, edit: Any, item: dict) -> None:
+        insert_text_format = item.get("insertTextFormat")
 
+        text_edit = item.get('textEdit')
+        if text_edit:
+            # insert the removed command completion item prefix
+            # so we don't have to calculate the offset for the textEdit range
+            for sel in self.view.sel():
+                self.view.insert(edit, sel.begin(), CompletionHandler.last_prefix)
 
-last_text_command = None
+            new_text = text_edit.get('newText')
 
+            range = Range.from_lsp(text_edit['range'])
+            edit_region = range_to_region(range, self.view)
 
-class CompletionHelper(sublime_plugin.EventListener):
-    def on_text_command(self, view: sublime.View, command_name: str, args: Optional[Any]) -> None:
-        global last_text_command
-        last_text_command = command_name
+            self.view.erase(edit, edit_region)
```

Comment:
Let's try to think of a math solution for this, so that we can replace an `erase`+`insert` operation with a single `replace` operation. If the completion is a snippet, then we are forced to use `erase` I think. But in the case of a plaintext completion, I hope we can do it with a single `replace`!

Reply:
We can use replace :)